File: Source/OldCode/Archive/2004 TVA Code Library/Source/EE/Phasor/IEEE1344/CommandFrame.vb
'*******************************************************************************************************
'  CommandFrame.vb - IEEE1344 Command Frame
'  Copyright © 2004 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2003
'  Primary Developer: James R Carroll, System Analyst [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  01/14/2005 - James R Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports TVA.Interop
Imports TVA.Shared.Bit
Imports TVA.Shared.DateTime
Imports TVA.Compression.Common

Namespace EE.Phasor.IEEE1344

    ' This class represents a command frame that can be sent to a PMU to elicit a desired response.  ABB PMU's won't begin
    ' a data broadcast until a command has been sent to "turn on" the real-time stream.
    Public Class CommandFrame

        Public Const FrameLength As Integer = 18

        Private m_timetag As Unix.TimeTag
        Private m_pmuIDCode As Int64
        Private m_command As PMUCommand

        ' Use this contructor to send a command to a PMU
        Public Sub New(ByVal pmuIDCode As Int64, ByVal command As PMUCommand)

            m_pmuIDCode = pmuIDCode
            m_command = command
            m_timetag = New Unix.TimeTag(DateTime.Now)

        End Sub

        ' Use this constuctor to receive a command (i.e., your code is acting as a PMU)
        Public Sub New(ByVal binaryImage As Byte(), ByVal startIndex As Integer)

            If binaryImage Is Nothing Then
                Throw New ArgumentNullException("BinaryImage was null - could not create command frame")
            ElseIf binaryImage.Length - startIndex < FrameLength Then
                Throw New ArgumentException("BinaryImage size from startIndex is too small - could not create command frame")
            Else
                m_timetag = New Unix.TimeTag(Convert.ToDouble(EndianOrder.ReverseToInt32(binaryImage, startIndex)))
                m_pmuIDCode = EndianOrder.ReverseToInt64(binaryImage, startIndex + 4)
                m_command = EndianOrder.ReverseToInt16(binaryImage, startIndex + 12)

                ' Validate buffer check sum
                If EndianOrder.ReverseToInt16(binaryImage, startIndex + FrameLength - 2) <> CRC16(-1, binaryImage, startIndex, FrameLength - 2) Then _
                    Throw New ArgumentException("Invalid buffer image detected - CRC16 of command frame did not match")
            End If

        End Sub

        Public Property TimeTag() As Unix.TimeTag
            Get
                Return m_timetag
            End Get
            Set(ByVal Value As Unix.TimeTag)
                m_timetag = Value
            End Set
        End Property

        Public Property PMUIDCode() As Int64
            Get
                Return m_pmuIDCode
            End Get
            Set(ByVal Value As Int64)
                m_pmuIDCode = Value
            End Set
        End Property

        Public Property Command() As PMUCommand
            Get
                Return m_command
            End Get
            Set(ByVal Value As PMUCommand)
                m_command = Value
            End Set
        End Property

        Public ReadOnly Property BinaryImage() As Byte()
            Get
                Dim buffer As Byte() = Array.CreateInstance(GetType(Byte), FrameLength)

                ' TODO: MODIFIED FOR PC37.118 TEST - RESTORE FOR REAL IEEE1344 CODE...
                buffer(0) = &HAA
                buffer(1) = Bit0 Or Bit6
                EndianOrder.SwapCopyBytes(Convert.ToUInt16(FrameLength), buffer, 2)
                EndianOrder.SwapCopyBytes(Convert.ToUInt16(235), buffer, 4)
                EndianOrder.SwapCopyBytes(Convert.ToUInt32(m_timetag.Value), buffer, 6)
                EndianOrder.SwapCopyBytes(Convert.ToUInt32(0), buffer, 10)
                EndianOrder.SwapCopyBytes(Convert.ToInt16(m_command), buffer, 14)
                EndianOrder.SwapCopyBytes(CRC_CCITT(-1, buffer, 0, 16), buffer, 16)

                Return buffer
            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/SelFastMessage/CommandFrame.cs
//*******************************************************************************************************
//  CommandFrame.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  04/27/2009 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Security.Permissions;
using PCS.IO.Checksums;
using PCS.Parsing;

namespace PhasorProtocols.SelFastMessage
{
    /// <summary>
    /// Represents the SEL Fast Message implementation of a <see cref="ICommandFrame"/> that can be sent or received.
    /// </summary>
    /// <remarks>
    /// SEL Fast Message command frames are designed only to be sent to a device, not received from a device. As a result
    /// this frame does not implement <see cref="ISupportFrameImage{T}"/> for automated frame parsing. This class
    /// exposes a constructor that accepts a binary image in order to manually parse a command frame.
    /// </remarks>
    [Serializable()]
    public class CommandFrame : CommandFrameBase
    {
        #region [ Members ]

        // Fields
        private MessagePeriod m_messagePeriod;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="CommandFrame"/> from the given <paramref name="binaryImage"/>.
        /// </summary>
        /// <param name="binaryImage">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="binaryImage"/>.</param>
        /// <remarks>
        /// This constructor is used by a consumer to parse a received SEL Fast Message command frame. Typically
        /// command frames are sent to a device. This constructor would used if this code was being used
        /// inside of a phasor measurement device.
        /// </remarks>
        /// <exception cref="ArgumentOutOfRangeException"><paramref name="length"/> is not large enough to parse frame.</exception>
        public CommandFrame(byte[] binaryImage, int startIndex, int length)
            : base(new CommandCellCollection(0), PhasorProtocols.DeviceCommand.ReservedBits)
        {
            if (length < 16)
                throw new ArgumentOutOfRangeException("length");

            // Validate check-sum
            if (!ChecksumIsValid(binaryImage, startIndex))
                throw new InvalidOperationException("Invalid binary image detected - check sum of " + this.GetType().Name + " did not match");
            
            // Validate SEL Fast Message data image
            if (binaryImage[startIndex] != Common.HeaderByte1 || binaryImage[startIndex + 1] != Common.HeaderByte2)
                throw new InvalidOperationException("Bad data stream, expected header bytes 0xA546 as first bytes in SEL Fast Message command frame, got 0x" + binaryImage[startIndex].ToString("X").PadLeft(2, '0') + binaryImage[startIndex + 1].ToString("X").PadLeft(2, '0'));

            Command = (DeviceCommand)binaryImage[startIndex + 9];

            if (Command == DeviceCommand.EnableUnsolicitedMessages)
                m_messagePeriod = (MessagePeriod)EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex + 14);
            
            // Validate check-sum
            int sumLength = FrameSize - 2;

            if (EndianOrder.BigEndian.ToUInt16(binaryImage, startIndex + sumLength) != CalculateChecksum(binaryImage, startIndex, sumLength))
                throw new InvalidOperationException("Invalid binary image detected - check sum of " + this.GetType().Name + " did not match");
        }

        /// <summary>
        /// Creates a new <see cref="CommandFrame"/> from specified parameters.
        /// </summary>
        /// <param name="command">The <see cref="PhasorProtocols.DeviceCommand"/> for this <see cref="CommandFrame"/>.</param>
        /// <param name="messagePeriod">The desired <see cref="SelFastMessage.MessagePeriod"/> for SEL device connection.</param>
        /// <remarks>
        /// This constructor is used by a consumer to generate an SEL Fast Message command frame.
        /// </remarks>
        public CommandFrame(PhasorProtocols.DeviceCommand command, MessagePeriod messagePeriod)
            : base(new CommandCellCollection(0), command)
        {
            if (command != PhasorProtocols.DeviceCommand.EnableRealTimeData && command != PhasorProtocols.DeviceCommand.DisableRealTimeData)
                throw new ArgumentException("SEL Fast Message does not support " + command + " device command.", "command");

            m_messagePeriod = messagePeriod;
        }

        /// <summary>
        /// Creates a new <see cref="CommandFrame"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected CommandFrame(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize command frame
            m_messagePeriod = (MessagePeriod)info.GetValue("messagePeriod", typeof(MessagePeriod));
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets or sets the <see cref="DeviceCommand"/> for this <see cref="CommandFrame"/>.
        /// </summary>
        public new DeviceCommand Command
        {
            // Interpret SEL specific device command
            get
            {
                switch (base.Command)
                {
                    case PhasorProtocols.DeviceCommand.EnableRealTimeData:
                        return DeviceCommand.EnableUnsolicitedMessages;
                    case PhasorProtocols.DeviceCommand.DisableRealTimeData:
                        return DeviceCommand.DisableUnsolicitedMessages;
                    default:
                        return DeviceCommand.Undefined;
                }
            }
            set
            {
                switch (value)
                {
                    case DeviceCommand.EnableUnsolicitedMessages:
                        base.Command = PhasorProtocols.DeviceCommand.EnableRealTimeData;
                        break;
                    case DeviceCommand.DisableUnsolicitedMessages:
                        base.Command = PhasorProtocols.DeviceCommand.DisableRealTimeData;
                        break;
                    default:
                        base.Command = PhasorProtocols.DeviceCommand.ReservedBits;
                        break;
                }
            }
        }

        /// <summary>
        /// Gets or sets the desired message period for the SEL device.
        /// </summary>
        public MessagePeriod MessagePeriod
        {
            get
            {
                return m_messagePeriod;
            }
            set
            {
                m_messagePeriod = value;
            }
        }

        /// <summary>
        /// Gets frame size based on <see cref="Command"/>.
        /// </summary>
        public byte FrameSize
        {
            get
            {
                return (byte)(Command == DeviceCommand.EnableUnsolicitedMessages ? 18 : 16);
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="HeaderImage"/>.
        /// </summary>
        protected override int HeaderLength
        {
            get
            {
                return 9;
            }
        }

        /// <summary>
        /// Gets the binary header image of the <see cref="DataFrame"/> object.
        /// </summary>
        protected override byte[] HeaderImage
        {
            get
            {
                byte[] buffer = new byte[9];

                buffer[0] = Common.HeaderByte1;
                buffer[1] = Common.HeaderByte2;
                buffer[2] = FrameSize;

                return buffer;
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="BodyImage"/>.
        /// </summary>
        protected override int BodyLength
        {
            get
            {
                // Total frame size - header length - crc value length
                return FrameSize - HeaderLength - 2;
            }
        }

        /// <summary>
        /// Gets the binary body image of this <see cref="CommandFrame"/>.
        /// </summary>
        protected override byte[] BodyImage
        {
            get
            {
                byte[] buffer = new byte[BodyLength];

                buffer[0] = (byte)Command;
                buffer[1] = 0xC0;
                buffer[3] = 0x20;
                
                // Only add desired message rate for enable command
                if (Command == DeviceCommand.EnableUnsolicitedMessages)
                    EndianOrder.BigEndian.CopyBytes((ushort)m_messagePeriod, buffer, 5);

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="DataFrame"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Defined Message Period", (ushort)MessagePeriod + ": " + MessagePeriod);

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Calculates checksum of given <paramref name="buffer"/>.
        /// </summary>
        /// <param name="buffer">Buffer image over which to calculate checksum.</param>
        /// <param name="offset">Start index into <paramref name="buffer"/> to calculate checksum.</param>
        /// <param name="length">Length of data within <paramref name="buffer"/> to calculate checksum.</param>
        /// <returns>Checksum over specified portion of <paramref name="buffer"/>.</returns>
        protected override ushort CalculateChecksum(byte[] buffer, int offset, int length)
        {
            // SEL Fast Message uses CRC16 to calculate checksum for frames
            return buffer.Crc16Checksum(offset, length);
        }

        /// <summary>
        /// Appends checksum onto <paramref name="buffer"/> starting at <paramref name="startIndex"/>.
        /// </summary>
        /// <param name="buffer">Buffer image on which to append checksum.</param>
        /// <param name="startIndex">Index into <paramref name="buffer"/> where checksum should be appended.</param>
        /// <remarks>
        /// We override default implementation since SEL Fast Message implements check sum for frames in little-endian.
        /// </remarks>
        protected override void AppendChecksum(byte[] buffer, int startIndex)
        {
            EndianOrder.BigEndian.CopyBytes(CalculateChecksum(buffer, 0, startIndex), buffer, startIndex);
        }

        /// <summary>
        /// Determines if checksum in the <paramref name="buffer"/> is valid.
        /// </summary>
        /// <param name="buffer">Buffer image to validate.</param>
        /// <param name="startIndex">Start index into <paramref name="buffer"/> to perform checksum.</param>
        /// <returns>Flag that determines if checksum over <paramref name="buffer"/> is valid.</returns>
        /// <remarks>
        /// We override default implementation since SEL Fast Message implements check sum for frames in little-endian.
        /// </remarks>
        protected override bool ChecksumIsValid(byte[] buffer, int startIndex)
        {
            int sumLength = BinaryLength - 2;
            return EndianOrder.BigEndian.ToUInt16(buffer, startIndex + sumLength) == CalculateChecksum(buffer, startIndex, sumLength);
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize command frame
            info.AddValue("messagePeriod", m_messagePeriod, typeof(MessagePeriod));
        }

        #endregion
    }
}